File: admin/src/lib/escort.ts
import { AFTER_HOURS_FIXTURE_STATE_KEY } from "./constants";
import type { EscortZoneRule, ZoneKey, ZoneLabel } from "./types";
import { VISITOR_TYPES } from "./constants";

export const ZONE_KEYS: ZoneKey[] = [
  "reception",
  "admin",
  "classroom",
  "sports",
  "lab",
  "restricted",
  "parking",
];

export const ZONE_DEFAULT_LABELS: Record<ZoneKey, string> = {
  reception: "Reception / Lobby",
  admin: "Admin block",
  classroom: "Classroom wing",
  sports: "Sports / playground",
  lab: "Labs / IT",
  restricted: "Restricted (principal / accounts / stores / server)",
  parking: "Bus bay / parking",
};

const ESCORT_DEFAULTS: Record<string, { escortRequired: boolean; allowedZones: ZoneKey[] }> = {
  Vendor: { escortRequired: true, allowedZones: ["reception", "admin"] },
  Parent: { escortRequired: false, allowedZones: ["reception"] },
  Guest: { escortRequired: false, allowedZones: ["reception"] },
  Official: { escortRequired: false, allowedZones: ["reception", "admin"] },
  Alumni: { escortRequired: false, allowedZones: ["reception"] },
};

export function seedZones(): ZoneLabel[] {
  return ZONE_KEYS.map((key) => ({
    key,
    label: ZONE_DEFAULT_LABELS[key],
    schoolId: "SCH-DEMO-01",
    updatedByUserId: "U-ADMIN",
  }));
}

export function seedEscortRules(): EscortZoneRule[] {
  return VISITOR_TYPES.map((visitorType) => {
    const def = ESCORT_DEFAULTS[visitorType] || ESCORT_DEFAULTS.Guest;
    return {
      schoolId: "SCH-DEMO-01",
      visitorType,
      escortRequired: def.escortRequired,
      allowedZones: [...def.allowedZones],
      updatedByUserId: "U-ADMIN",
    };
  });
}

export function zoneLabelMap(zones: ZoneLabel[]): Record<string, string> {
  const map: Record<string, string> = {};
  zones.forEach((z) => {
    map[z.key] = z.label;
  });
  ZONE_KEYS.forEach((key) => {
    if (!map[key]) map[key] = ZONE_DEFAULT_LABELS[key];
  });
  return map;
}

export function formatAllowedZones(keys: string[] | undefined, zones?: ZoneLabel[]): string {
  if (!keys || !keys.length) return "—";
  const labels = zoneLabelMap(zones || []);
  return keys.map((k) => labels[k] || k).join(", ");
}

export function escortCell(required?: boolean, name?: string | null, waived?: boolean): string {
  if (waived) return "Waived";
  if (!required) return "No";
  return name ? name : "Required";
}

const ESCORT_STATE_KEY = `${AFTER_HOURS_FIXTURE_STATE_KEY}-escort`;

export interface EscortFixtureSession {
  zones: ZoneLabel[];
  rules: EscortZoneRule[];
}

let session: EscortFixtureSession | null = null;

function clone<T>(v: T): T {
  return JSON.parse(JSON.stringify(v)) as T;
}

export function getEscortFixtureSession(
  zones?: ZoneLabel[],
  rules?: EscortZoneRule[],
): EscortFixtureSession {
  if (session) return session;
  try {
    const raw = sessionStorage.getItem(ESCORT_STATE_KEY);
    if (raw) {
      session = JSON.parse(raw) as EscortFixtureSession;
      return session;
    }
  } catch {
    /* ignore */
  }
  session = {
    zones: clone(zones && zones.length ? zones : seedZones()),
    rules: clone(rules && rules.length ? rules : seedEscortRules()),
  };
  return session;
}

function persist(): void {
  if (!session) return;
  try {
    sessionStorage.setItem(ESCORT_STATE_KEY, JSON.stringify(session));
  } catch {
    /* ignore */
  }
}

export function saveZonesLocal(zones: ZoneLabel[]): ZoneLabel[] {
  const sess = getEscortFixtureSession();
  sess.zones = zones.map((z) => ({ ...z, updatedAt: new Date().toISOString() }));
  persist();
  return clone(sess.zones);
}

export function saveRulesLocal(rules: EscortZoneRule[]): EscortZoneRule[] {
  const sess = getEscortFixtureSession();
  sess.rules = rules.map((r) => ({
    ...r,
    escortRequired: r.allowedZones.includes("restricted") ? true : r.escortRequired,
    updatedAt: new Date().toISOString(),
  }));
  persist();
  return clone(sess.rules);
}

File: app/escort.py
"""Escort / zone rules (Hub B4, AC-B4). Stamp at visit create; no geo-fence."""
from __future__ import annotations

from typing import Optional

from app.errors import AppError
from app import store

ZONE_KEYS = (
    "reception",
    "admin",
    "classroom",
    "sports",
    "lab",
    "restricted",
    "parking",
)

DEFAULT_ZONE_LABELS = {
    "reception": "Reception / Lobby",
    "admin": "Admin block",
    "classroom": "Classroom wing",
    "sports": "Sports / playground",
    "lab": "Labs / IT",
    "restricted": "Restricted (principal / accounts / stores / server)",
    "parking": "Bus bay / parking",
}

VISITOR_TYPES = ("Parent", "Vendor", "Guest", "Official", "Alumni")

# Hub-LOCKED new-school defaults
DEFAULT_ESCORT_RULES: dict[str, dict] = {
    "Vendor": {"escortRequired": True, "allowedZones": ["reception", "admin"]},
    "Parent": {"escortRequired": False, "allowedZones": ["reception"]},
    "Guest": {"escortRequired": False, "allowedZones": ["reception"]},
    "Official": {"escortRequired": False, "allowedZones": ["reception", "admin"]},
    "Alumni": {"escortRequired": False, "allowedZones": ["reception"]},
}

CONTRACTOR_ALIASES = frozenset({"contractor", "Contractor"})


def map_visitor_type(raw: Optional[str]) -> str:
    """Contractor maps to Vendor rules (B4-E6)."""
    if raw is None:
        return "Guest"
    text = str(raw).strip()
    if text.lower() == "contractor":
        return "Vendor"
    return text


def normalize_zones(keys: list) -> list[str]:
    seen: list[str] = []
    unknown: list[str] = []
    for item in keys or []:
        key = item.value if hasattr(item, "value") else str(item).strip()
        if key not in ZONE_KEYS:
            unknown.append(key)
            continue
        if key not in seen:
            seen.append(key)
    if unknown:
        raise AppError(
            "VALIDATION",
            f"Unknown zone key(s): {', '.join(unknown)} — starter enum is fixed",
            400,
            {"unknown": unknown, "allowed": list(ZONE_KEYS)},
        )
    return seen


def apply_restricted_force(escort_required: bool, allowed_zones: list[str]) -> bool:
    if "restricted" in allowed_zones:
        return True
    return bool(escort_required)


def default_rule(visitor_type: str) -> dict:
    vt = map_visitor_type(visitor_type)
    base = DEFAULT_ESCORT_RULES.get(vt) or DEFAULT_ESCORT_RULES["Guest"]
    zones = list(base["allowedZones"])
    return {
        "visitorType": vt,
        "escortRequired": apply_restricted_force(base["escortRequired"], zones),
        "allowedZones": zones,
    }


def get_rule(school_id: str, visitor_type: str) -> dict:
    vt = map_visitor_type(visitor_type)
    row = store.get_escort_rule(school_id, vt)
    if not row:
        rule = default_rule(vt)
        rule["schoolId"] = school_id
        return rule
    zones = normalize_zones(row.get("allowedZones") or [])
    return {
        "schoolId": school_id,
        "visitorType": vt,
        "escortRequired": apply_restricted_force(row.get("escortRequired"), zones),
        "allowedZones": zones,
        "updatedByUserId": row.get("updatedByUserId"),
        "updatedAt": row.get("updatedAt"),
    }


def stamp_escort_zones(visit: dict) -> dict:
    """Stamp sticky escortRequired + allowedZones from EscortZoneRule (by visitorType)."""
    vt = map_visitor_type(visit.get("visitorType"))
    visit["visitorType"] = vt
    rule = get_rule(visit["schoolId"], vt)
    visit["escortRequired"] = bool(rule["escortRequired"])
    visit["allowedZones"] = list(rule["allowedZones"])
    visit.setdefault("escortStaffId", None)
    visit.setdefault("escortSuggestedByHost", None)
    visit.setdefault("escortWaived", False)
    visit.setdefault("escortWaiveReason", None)
    visit.setdefault("escortClearedAt", None)
    return visit


def escort_name(visit: dict) -> Optional[str]:
    sid = visit.get("escortStaffId")
    if not sid:
        return None
    staff = store.get_staff(sid)
    return staff["name"] if staff else None


def allowed_zone_labels(school_id: str, keys: list[str]) -> list[str]:
    labels: list[str] = []
    for key in keys or []:
        row = store.get_zone(school_id, key)
        labels.append(row["label"] if row else DEFAULT_ZONE_LABELS.get(key, key))
    return labels


def escort_blocks_pass(visit: dict) -> bool:
    """AC-B4e: escortRequired && !escortStaffId && !escortWaived."""
    return bool(
        visit.get("escortRequired")
        and not visit.get("escortStaffId")
        and not visit.get("escortWaived")
    )


def assert_escort_ready(visit: dict, action: str = "check-in") -> None:
    if escort_blocks_pass(visit):
        raise AppError(
            "ESCORT_REQUIRED",
            f"Escort required — assign escort staff before {action}",
            403,
            {
                "escortRequired": True,
                "escortStaffId": visit.get("escortStaffId"),
                "escortWaived": bool(visit.get("escortWaived")),
                "allowedZones": list(visit.get("allowedZones") or []),
            },
        )


def seed_school_defaults(school_id: str, updated_by: Optional[str] = None) -> None:
    """Idempotent starter zones + escort rules for a school."""
    ts = None
    for key, label in DEFAULT_ZONE_LABELS.items():
        existing = store.get_zone(school_id, key)
        if existing:
            continue
        store.put_zone(
            {
                "key": key,
                "label": label,
                "schoolId": school_id,
                "updatedByUserId": updated_by,
                "updatedAt": ts,
            }
        )
    for vt in VISITOR_TYPES:
        existing = store.get_escort_rule(school_id, vt)
        if existing:
            continue
        rule = default_rule(vt)
        store.put_escort_rule(
            {
                "schoolId": school_id,
                "visitorType": vt,
                "escortRequired": rule["escortRequired"],
                "allowedZones": rule["allowedZones"],
                "updatedByUserId": updated_by,
                "updatedAt": ts,
            }
        )
